File: python/src/patterns/creational/singleton/double_check_lock_singleton.py
import threading


class DoubleCheckLockSingleton:
    _instance = None
    _lock = threading.Lock()

    def __init__(self) -> None:
        # prevent multiple instantiation by raising an exception
        if DoubleCheckLockSingleton._instance is not None:
            raise Exception("Cannot instantiate directly.")
        # init logic here

    @staticmethod
    def get_instance() -> "DoubleCheckLockSingleton":
        # only create once (the first time this method is called)
        # first check -> not synchronized
        if DoubleCheckLockSingleton._instance is None:
            # thread lock to allow one thread at a time
            with DoubleCheckLockSingleton._lock:
                # double check -> synchronized
                if DoubleCheckLockSingleton._instance is None:
                    DoubleCheckLockSingleton._instance = DoubleCheckLockSingleton()
        return DoubleCheckLockSingleton._instance

File: java/src/patterns/creational/singleton/DoubleCheckLockSingleton.java
package patterns.creational.singleton;

class DoubleCheckLockSingleton {
    // single static instance
    // volatile keyword for immediate visibility of changes to all threads
    private static volatile DoubleCheckLockSingleton instance;

    // private constructor prevents object instantiation
    private DoubleCheckLockSingleton() {
        // init logic here
    }

    public static DoubleCheckLockSingleton getInstance() {
        // only create once (the first time this method is called)
        // first check -> not synchronized
        if (instance == null) {
            // thread lock to allow one thread at a time
            synchronized (DoubleCheckLockSingleton.class) {
                // double check -> synchronized
                if (instance == null) {
                    instance = new DoubleCheckLockSingleton();
                }
            }
        }
        return instance;
    }
}
